zcash_client_backend/src/wallet.rs

```diff
@@ -297,27 +308,72 @@ impl<NoteRef> ReceivedNote<NoteRef> {
     pub fn output_index(&self) -> u16 {
         self.output_index
     }
-    pub fn note(&self) -> &Note {
+    pub fn note(&self) -> &NoteT {
         &self.note
     }
-    pub fn value(&self) -> NonNegativeAmount {
-        self.note.value()
-    }
     pub fn spending_key_scope(&self) -> Scope {
         self.spending_key_scope
     }
     pub fn note_commitment_tree_position(&self) -> Position {
         self.note_commitment_tree_position
     }
+
+    /// Applies the given function to the `note` field of this ReceivedNote and returns
+    /// `None` if that function returns `None`, or otherwise a `Some` containing
+    /// a `ReceivedNote` with its `note` field swapped out for the result of the function.
+    ///
+    /// The name `traverse` refers to the general operation that has the Haskell type
+    /// `Applicative f => (a -> f b) -> t a -> f (t b)`, that this method specializes
+    /// with `ReceivedNote<NoteRef, _>` for `t` and `Option<_>` for `f`.
+    pub fn traverse_opt<B>(
+        self,
+        f: impl FnOnce(NoteT) -> Option<B>,
+    ) -> Option<ReceivedNote<NoteRef, B>> {
+        f(self.note).map(|n0| ReceivedNote {
+            note_id: self.note_id,
+            txid: self.txid,
+            output_index: self.output_index,
+            note: n0,
+            spending_key_scope: self.spending_key_scope,
+            note_commitment_tree_position: self.note_commitment_tree_position,
+        })
+    }
+}
+
+impl<NoteRef> ReceivedNote<NoteRef, Note> {
+    pub fn protocol(&self) -> ShieldedProtocol {
```

Comment:
Document.

Reply:
Actually this is redundant; might as well just use `.note().protocol()`